File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.java

class Solution {
    public int largestRectangleArea(int[] arr) {
        int n=arr.length;
        int[] nse= new int[n];
        int[] pse= new int[n];

        Arrays.fill(nse,n);
        Arrays.fill(pse,-1);

        ArrayDeque<Integer> st = new ArrayDeque<>();
        st.add(n-1);
        for(int i=n-2;i>=0;i--){
            while(!st.isEmpty() && arr[st.peek()]>=arr[i]){
                st.pop();
            }
            if(!st.isEmpty()){
                nse[i] = st.peek();
            }
            st.push(i);
        }

        st.clear();

        st.add(0);
        for(int i=1;i<n;i++){
            while(!st.isEmpty() && arr[st.peek()]>=arr[i]){
                st.pop();
            }
            if(!st.isEmpty()){
                pse[i] = st.peek();
            }
            st.push(i);
        }

        System.out.println(Arrays.toString(nse));
        System.out.println(Arrays.toString(pse));

       int max=0;
        for(int i=0;i<n;i++){
                int width=nse[i]-pse[i]-1;
                int ans=arr[i]*width;
                max=Math.max(ans,max);
            }
        return max;

    }
}

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.cpp
class Solution {
public:
    vector<int> NSE(vector<int>& arr){
        int n=arr.size();
        vector<int> nse(n);
        stack<int>st;
        //st.push(n);

        for(int i=n-1;i>=0;i--){
            while(!st.empty() && arr[st.top()]>arr[i]){
                st.pop();
            }
            if(st.empty()){
                nse[i]=n;
            }
            else {
                nse[i]=st.top();
            }
            st.push(i);
        }
        return nse;
    }    
public:
    vector<int> PSEE(vector<int>& arr){
        int n=arr.size();
        vector<int> pse(n);
        stack<int>st;

        for(int i=0;i<n;i++){
            while(!st.empty() && arr[st.top()]>=arr[i]){
                st.pop();
            }
            if(st.empty()){
                pse[i]=-1;
            }
            else if(!st.empty()){
                pse[i]=st.top();
            }
            st.push(i);
        }
        return pse;
    }    
public:
    int largestRectangleArea(vector<int>& heights) {
        vector<int> nse=NSE(heights);
        vector<int> pse=PSEE(heights);
        int max1=0;
        for(int i=0;i<heights.size();i++){
            int dif=abs(pse[i]-nse[i])-1;
            int val=dif*heights[i];
            max1=max(max1,val);
        }
        return max1;
    }
};